### PR TITLE
Configuration refactor

### PR DESCRIPTION
**Pull request by dlemures**
We have 2 different setting variables: runtime checks and reflection usage.

However, we currently have 3 configurations that are not very clear:

Configuration | Runtime checks | Reflection
------------ | ------------- | ------------
Development | X | X
Production |  | X
Reflection free |  | 

Moreover, we might want to avoid reflection (using Registries), but have runtime checks for debug builds.

So, this PR is a refactor from:
```java
Configuration.development()
Configuration.production()
Configuration.reflectionFree()
```

to:
```java
Configuration.forDevelopment()  // Default to reflection
Configuration.forProduction()  // Default to reflection
Configuration.forDevelopment().enableReflection()
Configuration.forDevelopment().disableReflection()
Configuration.forProduction().enableReflection()
Configuration.forProduction().disableReflection()
```

Where development means to include runtime checks, and production the opposite.
This way both variables are totally independent.

By default, production (no runtime checks) and reflection are used, the same as now.
Also the same style is kept.

### CODE REVIEW

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ * Strategy pattern that allows to change various behaviors of Toothpick.
+ * The default configuration is {@link #production()} and {@link #reflection()}.
+ * A custom configuration can be created and used by toothpick,
+ * it is even possible to use a composition of the built-in configurations.
+ */
+public class Configuration {
+
+  public static Configuration instance = new Configuration();
+
+  private ReflectionConfiguration reflectionConfiguration = new ReflectionOnConfiguration();
```
Comment (mmadev) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: what is the reason we don't use reflection free by default? so TP run faster.

Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: It is a design decision we took. We can discuss it anyway.

If we use the reflectionfree configuration, some extra steps are needed (setting up registries, sending some annotation processing parameters).

So the default one is the one that will just run with no settings.

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+  /**
+   * Set reflection mode.
+   * It is slower than {@link #reflectionFree()} but it does not
+   * need any additional setup of the annotation processors.
+   * It can be used in production or development.
+   *
+   * @return a configuration set up to use reflection.
+   */
+  public Configuration reflection() {
```
Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: IMO something like ```enableReflection()``` would be more clear to understand.

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+  /**
+   * Set reflection free mode.
+   * It is faster than {@link #reflection()} but it needs
+   * some additional setup of the annotation processors.
+   * It can be used in production or development.
+   *
+   * @return an optimized reflection free configuration.
+   */
+  public Configuration reflectionFree() {
```
Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: IMO something like ```disableReflection()``` would be more clear to understand.

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+   * The checks performed are:
+   * <ul>
+   * <li>cycle detection: check that not 2 classes depend on each other. Note that if of them uses a Lazy instance
+   * of the other or a Producer, then there is no such cycle.</li>
+   * <li>illegal binding detection: check no scope annotated class is used as the target of a binding.</li>
+   * </ul>
+   *
+   * @return a development configuration.
+   */
+  public static Configuration development() {
```
Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: What about ```Configuration.forDevelopment()```?

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  /**
+   * Performs no runtime checks. This configuration
+   * is faster than {@link #development()}.
+   * It can be used in production.
+   *
+   * @return a production configuration.
+   */
+  public static Configuration production() {
```
Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: What about ```Configuration.forProduction()```?

File: toothpick-runtime/src/main/java/toothpick/configuration/ReflectionOffConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package toothpick.configuration;
+
+import toothpick.Factory;
+import toothpick.MemberInjector;
+import toothpick.registries.FactoryRegistryLocator;
+import toothpick.registries.MemberInjectorRegistryLocator;
+
+class ReflectionOffConfiguration implements ReflectionConfiguration {
```
Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/ReflectionOffConfiguration.java]: ```NoReflectionConfiguration``` ?

Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/ReflectionOffConfiguration.java]: ```NoReflectionConfiguration``` ?

File: toothpick-runtime/src/main/java/toothpick/configuration/ReflectionOnConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package toothpick.configuration;
+
+import toothpick.Factory;
+import toothpick.MemberInjector;
+import toothpick.registries.FactoryRegistryLocator;
+import toothpick.registries.MemberInjectorRegistryLocator;
+
+class ReflectionOnConfiguration implements ReflectionConfiguration {
```
Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/ReflectionOnConfiguration.java]: ```DefaultReflectionConfiguration``` ?

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import toothpick.MemberInjector;
+import toothpick.config.Binding;
+
+/**
+ * Strategy pattern that allows to change various behaviors of Toothpick.
+ * The default configuration is {@link #production()} and {@link #reflection()}.
+ * A custom configuration can be created and used by toothpick,
+ * it is even possible to use a composition of the built-in configurations.
+ */
+public class Configuration {
```
Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: final

Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: Implement both interfaces ```RuntimeConfiguration``` and ```ReflectionConfiguration```

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java (hunk 1 above)
Comment (mmadev) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: I guess you already have documents somewhere. Just make sure other users aware of this option, so they can have the speed benefits.

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java (hunk 2 above)
Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: done

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java (hunk 3 above)
Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: done

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java (hunk 4 above)
Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: actually it is not configuration for development. It is returning development like:
```java
Toothpick.setConfiguration(Configuration.development())
```

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java (hunk 5 above)
Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: same

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java (hunk 8 above)
Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: good point

Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: regarding the final, the user should be able to extend Configuration and provide a custom one.

File: toothpick-runtime/src/main/java/toothpick/configuration/ReflectionOnConfiguration.java (hunk 7 above)
Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/ReflectionOnConfiguration.java]: Maybe we can discuss this namings later.

I'd prefer to use:
IReflectionConfiguration
NoReflectionConfiguration
ReflectionConfiguration

But we don't use I prefix for interfaces right now. We could wait for @stephanenicolas to be back and discuss about it.

I'd prefer not to use Default... As an implementation doesn't know about what is default and say nothing about what it does.

File: toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java (hunk 1 above)
Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: Yes it is stated.
But probably you are right and we could add more info about it as it is one of the main gains of TP.

Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/Configuration.java]: Issue to track it: https://github.com/stephanenicolas/toothpick/issues/141

File: toothpick-runtime/src/main/java/toothpick/configuration/RuntimeCheckOnConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      if (annotation.annotationType().isAnnotationPresent(javax.inject.Scope.class)) {
+        throw new IllegalBindingException(format("Class %s cannot be bound. It has an scope annotation", clazz.getName()));
+      }
+    }
+  }
+
+  @Override
+  public void checkCyclesStart(Class clazz, String name) {
+    final Pair pair = new Pair(clazz, name);
+    final LinkedHashSet<Pair> linkedHashSet = cycleDetectionStack.get();
```
Comment (dlemures) [inline toothpick-runtime/src/main/java/toothpick/configuration/RuntimeCheckOnConfiguration.java]: Updated to use Pairs so name is taken into account when checking cycles

File: toothpick-runtime/src/main/java/toothpick/configuration/ConfigurationHolder.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package toothpick.configuration;
+
+/**
+ * Strategy pattern that allows to change various behaviors of Toothpick.
```
Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/ConfigurationHolder.java]: Update the documentation here.

File: toothpick-runtime/src/main/java/toothpick/configuration/ConfigurationHolder.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+public class ConfigurationHolder {
+
+  public static Configuration configuration = new Configuration();
+
+  /**
+   * Allows to pass custom configurations.
+   *
+   * @param configuration the configuration to use
+   */
+  public static void setConfiguration(Configuration configuration) {
```
Comment (saguinav) [inline toothpick-runtime/src/main/java/toothpick/configuration/ConfigurationHolder.java]: I think that for now it's fine to get rid off the setter too.

**Comment by saguinav** [comment]
LGTM

**Pull request merged**